### ISSUE TITLE
JSDoc @template in generic function results in type any instead

### ISSUE THREAD

**Issue by dominikschreiber** [Microsoft/TypeScript#30541]
**TypeScript Version:**  3.4.0-dev.20190322

**Search Terms:** JSDoc @template

**Code**

```ts
/**
 * @template T
 * @param {T[]} arr
 * @param {function(T): boolean} pred
 * @return {T[]}
 */
function filter(arr, pred) {
    return arr.length === 0
        ? []
        : (pred(arr[0]) ? [arr[0]] : []).concat(filter(arr.slice(1), pred));
}


function tsfilter<T>(arr: T[], pred: (e: T) => boolean): T[] {
    return arr.length === 0
        ? []
        : (pred(arr[0]) ? [arr[0]] : []).concat(filter(arr.slice(1), pred));
}

// no typeerror :(
alert('filter: ' + JSON.stringify(filter([1, 2, 3, 'A'], _ => _ % 2 == 0)))
// typeerror
alert('tsfilter: ' + JSON.stringify(tsfilter([1, 2, 3, 'A'], _ => _ % 2 == 0)))
```

**Expected behavior:**

The jsdoc-typed `filter` method should use the same types as the ts-typed `tsfilter` method, i.e. it should throw a type error as well.

**Actual behavior:**

The jsdoc-typed `filter` method gets typed as
```ts
function filter(arr: any, pred: any): any
```
whereas the ts-typed `tsfilter` method gets typed as
```ts
function tsfilter<T>(arr: T[], pred: (T) => boolean): T[]
```

**Playground Link:** https://www.typescriptlang.org/play/#src=%2F**%0D%0A%20*%20%40template%20T%0D%0A%20*%20%40param%20%7BT%5B%5D%7D%20arr%0D%0A%20*%20%40param%20%7Bfunction(T)%3A%20boolean%7D%20pred%0D%0A%20*%20%40return%20%7BT%5B%5D%7D%0D%0A%20*%2F%0D%0Afunction%20filter(arr%2C%20pred)%20%7B%0D%0A%20%20%20%20return%20arr.length%20%3D%3D%3D%200%0D%0A%20%20%20%20%20%20%20%20%3F%20%5B%5D%0D%0A%20%20%20%20%20%20%20%20%3A%20(pred(arr%5B0%5D)%20%3F%20%5Barr%5B0%5D%5D%20%3A%20%5B%5D).concat(filter(arr.slice(1)%2C%20pred))%3B%0D%0A%7D%0D%0A%0D%0A%0D%0Afunction%20tsfilter%3CT%3E(arr%3A%20T%5B%5D%2C%20pred%3A%20(e%3A%20T)%20%3D%3E%20boolean)%3A%20T%5B%5D%20%7B%0D%0A%20%20%20%20return%20arr.length%20%3D%3D%3D%200%0D%0A%20%20%20%20%20%20%20%20%3F%20%5B%5D%0D%0A%20%20%20%20%20%20%20%20%3A%20(pred(arr%5B0%5D)%20%3F%20%5Barr%5B0%5D%5D%20%3A%20%5B%5D).concat(filter(arr.slice(1)%2C%20pred))%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20no%20typeerror%0D%0Aalert('filter%3A%20'%20%2B%20JSON.stringify(filter(%5B1%2C%202%2C%203%2C%20'A'%5D%2C%20_%20%3D%3E%20_%20%25%202%20%3D%3D%200)))%0D%0A%2F%2F%20typeerror%0D%0Aalert('tsfilter%3A%20'%20%2B%20JSON.stringify(tsfilter(%5B1%2C%202%2C%203%2C%20'A'%5D%2C%20_%20%3D%3E%20_%20%25%202%20%3D%3D%200)))

**Related Issues:** #1178

**Comment by ajafff** [comment]
JSDoc has no effect in .ts files (besides documentation).
If you move the declaration of `filter` to a .js file it should work as expected.

You should also consider turning on `noImplicitAny` as it will show you untyped parameters.

**Comment by dominikschreiber** [comment]
Thanks for pointing that out, @ajafff. I dug a little deeper and found that the problem really was on my side:

I in fact experienced the behavior in a js file, with the intellisense provided by Visual Studio Code. More specifically, I experienced it in two js files – one file defined the function, another file consumed it (both es3 sources). In the consuming file, I got no intellisense on the dynamic type `T` (the generic), it was always `any`. Turns out I did not have `checkJs` enabled (neither via a `jsconfig.json` nor via `// @ts-check`). Turning it on brings up exactly the expected errors and intellisense.

So: all my bad, sorry for bothering.
